KAFKA-19858: Set default min.insync.replicas=2 for __remote_log_metadata topic to prevent data loss (KIP-1235)

jiafu1115 opened the pull request:
[KIP-1235](https://cwiki.apache.org/confluence/x/yommFw)

The __remote_log_metadata internal topic currently lacks a configurable
min.insync.replicas setting, relying on the broker-level default
(typically 1) when the factor is 3 as default. This creates a data loss
risk in production environments, as writes may be acknowledged by only a
single replica.

What's more, this is inconsistent with some other cases such as
__transaction_state, which explicitly sets min.insync.replicas=2 via the
transaction.state.log.min.isr broker configuration. Both topics store
critical metadata and should have equivalent durability guarantees.

Note:
```
public static final String
REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP =
"rlmm.config.remote.log.metadata.topic.replication.factor";
public static final short
DEFAULT_REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR = 3;
```

Reviewers: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>,
 Chia-Ping Tsai <chia7712@gmail.com>

jiafu1115 committed:
KAFKA-19858: Set default min.insync.replicas=2 for __remote_log_metadata topic to prevent data loss

chia7712 commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -46,13 +46,15 @@ public final class TopicBasedRemoteLogMetadataManagerConfig {
     public static final String REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP = "remote.log.metadata.topic.replication.factor";
     public static final String REMOTE_LOG_METADATA_TOPIC_PARTITIONS_PROP = "remote.log.metadata.topic.num.partitions";
     public static final String REMOTE_LOG_METADATA_TOPIC_RETENTION_MS_PROP = "remote.log.metadata.topic.retention.ms";
+    public static final String REMOTE_LOG_METADATA_TOPIC_MIN_ISR_PROP = "remote.log.metadata.topic.min.isr";
```
@jiafu1115 that is a good idea, but I think this new config still need a KIP. Would you mind opening a KIP for it?

jiafu1115 replied:
@chia7712 Sure. thanks for your reminder.

AndrewJSchofield replied:
Also, maybe if there's going to be a KIP in this area, we should take the opportunity to make `__remote_log_metadata` into an internal topic.

jiafu1115 replied:
@AndrewJSchofield ok. I will consider this and try to combine two things into one KIP. thanks.

jiafu1115 replied:
@AndrewJSchofield Hi. I checked the code about the "internal" topic and summarised the result as follows:

<img width="1014" height="711" alt="image" src="https://github.com/user-attachments/assets/d7bdd150-e255-448c-b967-799f399bbc42" />
It seems that __remote_log_metadata isn’t a purely internal topic, since it can be configured as a normal topic in the remote Kafka cluster — even though its name can’t be changed. And it is using the standard producer to write data instead of local append directly.

```
org.apache.kafka.server.log.remote.storage.RemoteLogManager#configAndWrapRlmmPlugin
private Plugin<RemoteLogMetadataManager> configAndWrapRlmmPlugin(RemoteLogMetadataManager rlmm) {
    final Map<String, Object> rlmmProps = new HashMap<>();
    endpoint.ifPresent(e -> {
        rlmmProps.put(REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX + "bootstrap.servers", e.host() + ":" + e.port());
        rlmmProps.put(REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX + "security.protocol", e.securityProtocol().name);
    });
    // update the remoteLogMetadataProps here to override endpoint config if any
    rlmmProps.putAll(rlmConfig.remoteLogMetadataManagerProps());
    // ...
}
```

Besides, the current producer clientId for __remote_log_metadata will conflict with the follow code will cause write's reject:

```
//KafkaApis.scala
val internalTopicsAllowed = request.header.clientId == "__admin_client"
```

So maybe we shouldn’t make it an “internal” topic in this KIP when we’re not sure it’s matched with current design, and it will also help to keep this KIP stay simple and clear without any concern. WDTY? 

also cc @chia7712 @kamalcph for more thoughts.

Thanks!

jiafu1115 replied:
@kamalcph one more question need your input. Thanks a lot. Do we have real use case for configure one remote kafka cluster to store the topic (though current code support it) ?  Maybe nobody will use this configure in production, right?

jiafu1115 replied:
I’ve thought of one method to ensure compatibility during rollout if we really want to add it to internal topics to get lots of benefits from internal topic.

```
  val internalTopicsAllowed = 
    request.header.clientId == "__admin_client" ||
    request.header.clientId.startsWith("__remote_log_metadata_client") //Add this line to avoid rejecting writes.
```
or 

```
  val internalTopicsAllowed = 
    request.header.clientId.startsWith("__") 
```
//Another benefit is that the producer client ID can be dynamically generated with an identifier later on.

    
note:
```
org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerConfig:
private static final String REMOTE_LOG_METADATA_CLIENT_PREFIX = "__remote_log_metadata_client";
clientIdPrefix = REMOTE_LOG_METADATA_CLIENT_PREFIX + "_" + props.get(BROKER_ID);
```

jiafu1115 replied:
I open another PR for example: move the topic to internal:
https://github.com/apache/kafka/pull/20822

AndrewJSchofield replied:
@jiafu1115 OK, thanks for investigating the internal topic question. I agree that it's best not to make it an internal topic, because of the pluggable nature of this topic's use. Keeping your KIP simple is fine with me.

jiafu1115 replied:
@AndrewJSchofield 
My pleasure! I also learn a lot from the investigating. I’ve already opened a PR(https://github.com/apache/kafka/pull/20822) for it, and it doesn’t need a KIP. We can continue the discussion there later. After all, the topic really feels like an internal one — making it internal could have some nice benefits.
For this PR, if you think the approach looks good, I can prepare a tiny KIP now for more discussion. Thanks!

AndrewJSchofield replied:
Yes, I think a separate tiny KIP for the internal topic question seems like a good way to get the community to decide whether it's a good thing. Your PR for it looks about right. Of course, we'd need the KIP to pass before taking the PR.

jiafu1115 replied:
@AndrewJSchofield ok. for this minsync change. it is time for me to create one KIP. right? thanks

AndrewJSchofield replied:
I think you need a KIP for min.insync change. If you want to pursue the internal topic question, you could open a second KIP. There's no real reason to tangle the two questions together. Up to you though.

kamalcph commented:
If min-isr is set to 2, then when two nodes goes down in the cluster, all the upload/delete operations won't make progress for those unavailable metadata partitions. It is good to leave it to 1 by default, the user can change the settings dynamically based on the replication-factor of the topic using `kafka-configs.sh` script.

jiafu1115 commented:
> If min-isr is set to 2, then when two nodes goes down in the cluster, all the upload/delete operations won't make progress for those unavailable metadata partitions. It is good to leave it to 1 by default, the user can change the settings dynamically based on the replication-factor of the topic using `kafka-configs.sh` script.

@kamalcph  

Thanks for you comments. I think it is one broker broken vs two brokers broken. I list some extra points for why we need to consider to change it：

**Metadata loss is more critical than unavailable**
With min.isr=1, Leader crash = permanent metadata loss → remote data becomes unreachable. Silent data loss is worse than temporary write blocking.

**Edge case vs. common risk**
Single node failure (common): min.isr=2 still works 
Two nodes down (rare): Write blocked 
Correct trade-off: Block writes in extreme cases rather than risk data loss

**Industry standard or best pratice**
Kafka's own metadata topics use min.isr=2:
__transaction_state → min.isr=2

**Secure by default**
Most users won't tune internal configs. Defaults must be safe. Users needing higher availability can explicitly lower it.

Thanks for your discussion!

jiafu1115 commented:
@AndrewJSchofield @chia7712 @kamalcph https://cwiki.apache.org/confluence/display/KAFKA/KIP-1235%3A+Correct+the+default+min.insync.replicas+to+2+for+the+__remote_log_metadata+topic the KIP created for discussion. thanks
cc  @TaiJuWu

jiafu1115 commented:
@AndrewJSchofield @chia7712 @kamalcph 
Hi: I’ve started the vote for this KIP. If you think it makes sense, I’d appreciate your support by voting so that we can move it forward, although it’s not urgent. Thanks a lot!

jiafu1115 committed:
add related test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
spotless check fix

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
fix integration test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Merge branch 'keepsynctotrunk' of https://github.com/jiafu1115/kafka into minsync

jiafu1115 committed:
add the missed metadataTopicMinIsr in toString

Signed-off-by: stroller <fujian1115@gmail.com>

jiafu1115 committed:
Merge remote-tracking branch 'origin/minsync' into minsync

jiafu1115 commented:
### Integration Test Result:

#### Test Case Core flow:

1. Deploy the code to one of brokers
2. Change the configure if need
rlmm.config.remote.log.metadata.topic.replication.factor
rlmm.config.remote.log.metadata.topic.min.isr
3. Delete the topic __remote_log_metadata
4. Restart the broker
5. Check the topic info with command:
/kafka-topics.sh --bootstrap-server _ip:port_ --describe  --topic  __remote_log_metadata   --command-config /etc/kafka/admin_config.properties

#### Test Result:
| Configure\Code| Trunk without this PR| Trunk with this PR |
|--------|--------|--------|
|Not Configure|Case1: factor: 3 + min.isr: 1|Case3: factor: 3 + min.isr: 2|
|Configure: rlmm.config.remote.log.metadata.topic.replication.factor=2 rlmm.config.remote.log.metadata.topic.min.isr=1|Case2: factor: 2 + min.isr:1|Case4: factor: 2 + min.isr: 1|

Case1 Resut: 
Topic: __remote_log_metadata    TopicId: 9XTIc4rYTXKHyzBrwu_cMw PartitionCount: 50      **ReplicationFactor: 3**    Configs: remote.storage.enable=false,**min.insync.replicas=1**,cleanup.policy=delete,segment.bytes=1073741824,retention.ms=-1,max.message.bytes=1073741824,preallocate=false,segment.index.bytes=10485760

Case2 Result:
Topic: __remote_log_metadata    TopicId: T_qtmAgkSeCnUu8-Bji3Mw PartitionCount: 50      **ReplicationFactor: 2**    Configs: remote.storage.enable=false,**min.insync.replicas=1**,cleanup.policy=delete,segment.bytes=1073741824,[retention.ms](https://retention.ms/)=-1,max.message.bytes=1073741824,preallocate=false,segment.index.bytes=10485760

Case3 Result:
Topic: __remote_log_metadata    TopicId: w5SjlIapSe-kmu5AGsoMrw PartitionCount: 50      **ReplicationFactor: 3**    Configs: remote.storage.enable=false,**min.insync.replicas=2**,cleanup.policy=delete,segment.bytes=1073741824,retention.ms=-1,max.message.bytes=1073741824,preallocate=false,segment.index.bytes=10485760

Case4 Result:
Topic: __remote_log_metadata    TopicId: h1ANnhjZS5C92AsrEA4Qwg PartitionCount: 50      **ReplicationFactor: 2**    Configs: remote.storage.enable=false,**min.insync.replicas=1**,cleanup.policy=delete,segment.bytes=1073741824,retention.ms=-1,max.message.bytes=1073741824,preallocate=false,segment.index.bytes=10485760

jiafu1115 commented:
@chia7712 @AndrewJSchofield @kamalcph cc @TaiJuWu 
Hi All:
The KIP has been accepted. Thank you for your vote.
I completed the tests today, and the results are posted above.
You can review them **at your convenience; it’s not urgent.**

Thanks a lot !

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+    public void testRemoteLogMetadataTopicMinIsr() throws ExecutionException, InterruptedException {
+        // Initialize the manager which will create the __remote_log_metadata topic
+        TopicBasedRemoteLogMetadataManager rlmm = topicBasedRlmm();
+        verifyRemoteLogMetadataTopicMinIsr(rlmm, (short) 2, "default value");
```
```suggestion
        verifyRemoteLogMetadataTopicMinIsr(remoteLogMetadataManager, (short) 2, "default value");
```

kamalcph replied:
Could you address this minor comment? Why do we want to re-initialize the RLMM?

jiafu1115 replied:
Hi @kamalcph 

The field TopicBasedRemoteLogMetadataManagerTest#remoteLogMetadataManager maybe null if the first running case is this one. So we can using topicBasedRlmm() similar with some other cases to make sure it must be initialized. If it is already initialized. it will return the existed one.

BTW: I remove all others cases and debug with the case if we only use remoteLogMetadataManager directly instead of topicBasedRlmm(). You can see it is null.
<img width="1047" height="277" alt="Image" src="https://github.com/user-attachments/assets/4f9126d9-4bd7-4c0a-b323-7b4a415c0f94" />

thanks a lot!

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+            verifyRemoteLogMetadataTopicMinIsr(customRlmm, customMinIsr, "custom value");
+        } finally {
+            customRlmm.close();
+        }
```
```suggestion
        try (TopicBasedRemoteLogMetadataManager customRlmm = RemoteLogMetadataManagerTestUtils.builder()
                .bootstrapServers(clusterInstance.bootstrapServers())
                .overrideRemoteLogMetadataManagerProps(overrideProps)
                .build()) {
            verifyRemoteLogMetadataTopicMinIsr(customRlmm, customMinIsr, "custom value");
        }
```

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+            
+            // Wait for the topic to be created
+            // RemoteLogMetadataManagerTestUtils uses 3 partitions for the metadata topic
+            clusterInstance.waitTopicCreation(metadataTopic, 3);
```
```suggestion
            clusterInstance.waitTopicCreation(metadataTopic, RemoteLogMetadataManagerTestUtils.METADATA_TOPIC_PARTITIONS_COUNT);
```

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+            DescribeConfigsResult describeResult = admin.describeConfigs(List.of(topicResource));
+            Config config = describeResult.all().get().get(topicResource);
+            
+            assertTrue(config != null, "Topic config should not be null");
```
Please replace assertTrue -> assertNotNull

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+            
+            assertTrue(config != null, "Topic config should not be null");
+            ConfigEntry minIsrEntry = config.get(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG);
+            assertTrue(minIsrEntry != null, "min.insync.replicas config should exist");
```
Please replace assertTrue -> assertNotNull

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -370,4 +375,58 @@ public void testInitializationFailure() throws IOException, InterruptedException
+
+    private void verifyRemoteLogMetadataTopicMinIsr(TopicBasedRemoteLogMetadataManager rlmm, 
+                                                     short expectedMinIsr, 
+                                                     String valueDescription) 
```
nit: align the parameter indentation.

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfigTest.java:
```diff
@@ -192,4 +194,20 @@ private void assertMaskedSensitiveConfigurations(String configString) {
+
+    @Test
+    public void testDefaultMinIsr() {
+        Map<String, Object> props = createValidConfigProps(new HashMap<>(), new HashMap<>(), new HashMap<>());
```
Instead of `new HashMap<>`, can it be replaced with `Collections.emptyMap`


```suggestion
        Map<String, Object> props = createValidConfigProps(Collections.emptyMap(), Collections.emptyMap(), Collections.emptyMap());
```

kamalcph commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfigTest.java:
```diff
@@ -192,4 +194,20 @@ private void assertMaskedSensitiveConfigurations(String configString) {
+
+    @Test
+    public void testCustomMinIsr() {
+        Map<String, Object> props = createValidConfigProps(new HashMap<>(), new HashMap<>(), new HashMap<>());
```
same as above: 


```suggestion
        Map<String, Object> props = createValidConfigProps(Collections.emptyMap(), Collections.emptyMap(), Collections.emptyMap());
```

kamalcph reviewed (commented):
Thanks for the PR! Left few minor comments to address. PTAL. 

Could you also update the upgrade notes to inform the user to change the `min.insync.replicas` value to 2 if the `__remote_log_metadata` topic already created in the cluster using `kafka-configs.sh` script?

jiafu1115 committed:
new HashMap<> can it be replaced with Collections.emptyMap

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

jiafu1115 committed:
new HashMap<> can be replaced with Collections.emptyMap

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

jiafu1115 committed:
Update storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

jiafu1115 committed:
avoid test resource leak

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

jiafu1115 committed:
Update storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

jiafu1115 committed:
fix unit test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
fix code style issue

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
fix code style issue

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Fix formatting in testRemoteLogMetadataTopicWithDefaultMinIsr

jiafu1115 committed:
refactor the unit test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Merge remote-tracking branch 'origin/minsync' into minsync

jiafu1115 committed:
refactor the unit test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
add the configure into upgrade.md

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 commented:
@kamalcph Thanks for the thorough review! I’ve addressed the comments and updated the changes. You can help to take another look when you’re free? Thanks!

kamalcph commented on docs/operations/tiered-storage.md:
```diff
@@ -89,9 +89,13 @@ After build successfully, there should be a `kafka-storage-x.x.x-test.jar` file
+    # For single broker cluster, set this to 1. Default is 3 for clusters with 3 or more brokers.
     rlmm.config.remote.log.metadata.topic.replication.factor=1
     
+    # The minimum number of replicas that must acknowledge a write to remote log metadata topic.
```
```suggestion
    # The minimum number of insync replicas that must acknowledge a write to remote log metadata topic.
```

jiafu1115 replied:
https://github.com/apache/kafka/pull/20811#discussion_r2649487221 the same reply for this. thanks

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -63,6 +65,7 @@ public final class TopicBasedRemoteLogMetadataManagerConfig {
             "Default: -1, that means unlimited. Users can configure this value based on their use cases. " +
             "To avoid any data loss, this value should be more than the maximum retention period of any topic enabled with " +
             "tiered storage in the cluster.";
+    public static final String REMOTE_LOG_METADATA_TOPIC_MIN_ISR_DOC = "The minimum number of replicas that must acknowledge a write to remote log metadata topic.";
```
```suggestion
    public static final String REMOTE_LOG_METADATA_TOPIC_MIN_ISR_DOC = "The minimum number of insync replicas that must acknowledge a write to remote log metadata topic.";
```

jiafu1115 replied:
@kamalcph 
Thanks for your review. I feel it might be slightly redundant to use ‘insync’ as a prefix.” , `acknowledge `is the expected behavior and `insync `is the result of behavior. So WDTY? just keep it unchanged so that keep the same style with TRANSACTIONS_TOPIC_MIN_ISR_DOC ? 

Note:  TRANSACTIONS_TOPIC_MIN_ISR_DOC
    public static final String TRANSACTIONS_TOPIC_MIN_ISR_DOC = "The minimum number of replicas that must acknowledge a write to transaction topic in order to be considered successful.";

kamalcph replied:
ack, we can keep it to same.

kamalcph reviewed (commented):
Overall LGTM, left few minor comments. PTAL.

jiafu1115 commented:
Hi @kamalcph 

For this comment: "Could you address this minor comment? Why do we want to re-initialize the RLMM?"
https://github.com/apache/kafka/pull/20811#discussion_r2649463809

The field TopicBasedRemoteLogMetadataManagerTest#remoteLogMetadataManager maybe null if the first running case is this one. So we can using topicBasedRlmm() similar with some other cases to make sure it must be initialized. If it is already initialized. it will return the existed one.

So using topicBasedRlmm() will make sure every case is dedicated and runnable one.

BTW: I remove all others cases and debug with the case if we only use remoteLogMetadataManager directly instead of topicBasedRlmm(). You can see it is null.
<img width="961" height="317" alt="image" src="https://github.com/user-attachments/assets/c45cfbf4-7804-4a32-81ce-e040a9516ce8" />

thanks a lot!

kamalcph reviewed (approved):
LGTM, thanks for the PR!

chia7712 commented on docs/getting-started/upgrade.md:
```diff
@@ -33,7 +33,7 @@ type: docs
 
   * Two new configs have been introduced: `group.coordinator.cached.buffer.max.bytes` and `share.coordinator.cached.buffer.max.bytes`. They allow the respective coordinators to set the maximum buffer size retained for reuse. For further details, please refer to [KIP-1196](https://cwiki.apache.org/confluence/x/hA5JFg). 
-
+  * The new config have been introduced: `rlmm.config.remote.log.metadata.topic.min.isr` with 2 as default value. You can update the value for the existed __remote_log_metadata topic via kafka-configs.sh if needed. For further details, please refer to [KIP-1235](https://cwiki.apache.org/confluence/x/yommFw).
```
Since `rlmm.config` is just a prefix, we should refer to `remote.log.metadata.topic.min.isr` instead. Also, for existing topics, users need to update `min.insync.replicas ` rather than `remote.log.metadata.topic.min.isr`.

jiafu1115 replied:
Hi @chia7712 
Thanks for your reminder. I checked the code and found the prefix can be configured by "RemoteLogManagerConfig#REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX_PROP". So it make sense to do the change.  I already done with change. You can help to review again. thanks!

chia7712 reviewed (commented):
@jiafu1115 thanks for this patch. overall LGTM. one trivial comment remains

jiafu1115 committed:
Fix config name and update min.isr description

Corrected the config name and updated the description for setting min.insync.replicas.
